subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/ExecuteActionsTaskExecuter.java

```diff
@@ -85,16 +98,78 @@ private GradleException executeActions(TaskInternal task, TaskStateInternal stat
             } finally {
                 task.getStandardOutputCapture().stop();
             }
+            actionNumber++;
         }
         return null;
     }
 
-    private void executeAction(TaskInternal task, ContextAwareTaskAction action, TaskExecutionContext context) {
+    private void executeAction(String displayName, final TaskInternal task, final ContextAwareTaskAction action, TaskExecutionContext context) {
         action.contextualise(context);
-        try {
-            action.execute(task);
-        } finally {
-            action.contextualise(null);
+        buildOperationExecutor.run(displayName, new Action<BuildOperationContext>() {
+            @Override
+            public void execute(BuildOperationContext buildOperationContext) {
+                Throwable actionFailure = null;
+                try {
+                    action.execute(task);
+                } catch (Throwable t) {
+                    actionFailure = t;
+                } finally {
+                    action.contextualise(null);
+                }
+
+                try {
```

Comment:
I feel like "complete a main action and wait for its async work to complete" will be a recurring pattern and should probably be extracted. For instance I can inject the `WorkerExecutor` into a piece of work and schedule more work (or at least I *should* be able to) and my main piece of work should then wait for the additionally scheduled work.